File: sdk/aas-verifier-py/src/aas_verifier/decoders/extension.py
"""extension-registry ``Extension`` account decoder — the mund-creanc-witer
obligation edge. Mirrors ``programs/extension-registry/src/state/extension.rs``
and the TS reference at ``sdk/aas-verifier-ts/src/decoders/extension.ts``
(either is authoritative; both track the same on-chain layout).

Unlike Shillbot's ``Task``, an ``Extension`` has no on-chain state enum: the
account exists only while the obligation is ACTIVE — ``attest_return_substance``
(success) and ``default_extension`` (slash) both CLOSE it, emitting the terminal
event. A present account is "active"; a closed one is terminal and un-attestable
(verifier ``account_closed``), exactly like a finalized Task.

This is a typed decoder + type helper for downstream consumers, NOT a drop-in
``ProtocolHandler``: that interface's ``DecodedAccount`` is Task-shaped, and the
verifier's step-5 field equality compares Task fields. Verifying an Extension
*attestation* end-to-end needs the verifier core generalized per ``account_kind``
(tracked separately).

Body offsets (after the 8-byte Anchor discriminator):

    off  size  field
      0    32  extender        (Pubkey)
     32    32  recipient       (Pubkey)
     64     1  extension_type  (u8)
     65     8  bond_lamports   (u64 LE)
     73     8  created_at      (i64 LE, unix seconds)
     81     1  bump
     82    16  _reserved       (skipped)

Body = 98 bytes; full account (with discriminator) = 106 = ``Extension::SPACE``.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import Optional

import base58

#: Anchor account name (-> discriminator ``sha256("account:Extension")[0:8]``).
EXTENSION_ACCOUNT_KIND = "Extension"

#: Minimum body length carrying every decoded field (through ``bump``).
EXTENSION_MIN_BODY_LEN = 82

#: A live Extension has no state byte — existence == active.
EXTENSION_ACTIVE_STATE = "active"

_EXTENSION_TYPE_NAMES = {
    0: "capability_validation",
    1: "mentorship",
    2: "sponsorship",
    3: "validation",
}


@dataclass(frozen=True)
class ExtensionRecord:
    extender: str
    recipient: str
    extension_type: int
    bond_lamports: int
    created_at: int
    bump: int


def decode_extension(data: bytes, account_kind: str) -> ExtensionRecord:
    """Decode an ``Extension`` account body (bytes AFTER the discriminator)."""
    if account_kind != EXTENSION_ACCOUNT_KIND:
        raise ValueError(
            f'decode_extension only handles account_kind="Extension", got "{account_kind}"'
        )
    if len(data) < EXTENSION_MIN_BODY_LEN:
        raise ValueError(
            f"Extension body too short: {len(data)} bytes "
            f"(need >= {EXTENSION_MIN_BODY_LEN})"
        )
    return ExtensionRecord(
        extender=base58.b58encode(data[0:32]).decode("ascii"),
        recipient=base58.b58encode(data[32:64]).decode("ascii"),
        extension_type=data[64],
        bond_lamports=struct.unpack_from("<Q", data, 65)[0],
        created_at=struct.unpack_from("<q", data, 73)[0],
        bump=data[81],
    )


def resolve_extension_type(extension_type: int) -> Optional[str]:
    """Extension-type discriminant -> name (constants.rs taxonomy). MVP records
    only CapabilityValidation (0). Returns ``None`` for unknown discriminants."""
    return _EXTENSION_TYPE_NAMES.get(extension_type)

File: sdk/client/src/vow/decoders/extension.ts
import { PublicKey } from "@solana/web3.js";

/**
 * extension-registry `Extension` account decoder — the mund-creanc-witer
 * obligation edge (extender vouches for recipient, backed by a bonded SOL
 * stake). Mirrors `programs/extension-registry/src/state/extension.rs` exactly.
 *
 * Unlike Shillbot's `Task`, an `Extension` has NO on-chain state enum: the
 * account exists only while the vouch is LIVE. Both terminal transitions
 * close it — `withdraw_extension` (extender reclaims bond) and
 * `default_extension` (slash to treasury) — emitting `ExtensionWithdrawn` /
 * `ExtensionDefaulted` events. So a present account is always "active"; a closed
 * one is terminal (the verifier returns `account_closed`), exactly mirroring a
 * finalized Task. The durable terminal history is the event stream, not an
 * account.
 *
 * NB: this exposes a typed decoder + discriminator/state helpers for downstream
 * consumers (the web-position indexer, an extension-aware verifier). It is NOT a
 * drop-in `ProtocolHandler`: that interface's `DecodedAccount` is Task-shaped
 * (task_id, composite_score, content hashes…), and `verify.ts` step 5 compares
 * those Task fields directly. Verifying an Extension *attestation* end-to-end
 * therefore needs the verifier's field-equality core generalized per
 * account_kind (and matching wire fields) — a deliberate change beyond this
 * decoder, tracked separately.
 *
 * Body offsets (after the 8-byte Anchor discriminator):
 *
 *   off  size  field
 *     0    32  extender        (Pubkey)
 *    32    32  recipient       (Pubkey)
 *    64     1  extension_type  (u8)
 *    65     8  bond_lamports   (u64 LE)
 *    73     8  created_at      (i64 LE, unix seconds)
 *    81     1  bump
 *    82    16  _reserved       (skipped)
 *
 * Body = 98 bytes; full account (with discriminator) = 106 = `Extension::SPACE`.
 */
export interface ExtensionRecord {
  extender: string;
  recipient: string;
  extension_type: number;
  bond_lamports: bigint;
  created_at: bigint;
  bump: number;
}

/** The Anchor account name (→ discriminator `sha256("account:Extension")[0..8]`). */
export const EXTENSION_ACCOUNT_KIND = "Extension";

/** Minimum body length carrying every decoded field (through `bump`). */
export const EXTENSION_MIN_BODY_LEN = 82;

/** A live Extension has no state byte — existence == active. */
export const EXTENSION_ACTIVE_STATE = "active";

/**
 * Decode an `Extension` account body (the bytes AFTER the 8-byte discriminator).
 * Throws on the wrong `accountKind` or a body too short for the decoded fields.
 */
export function decodeExtension(
  data: Uint8Array,
  accountKind: string
): ExtensionRecord {
  if (accountKind !== EXTENSION_ACCOUNT_KIND) {
    throw new Error(
      `decodeExtension only handles account_kind="Extension", got "${accountKind}"`
    );
  }
  if (data.length < EXTENSION_MIN_BODY_LEN) {
    throw new Error(
      `Extension body too short: ${data.length} bytes (need >= ${EXTENSION_MIN_BODY_LEN})`
    );
  }
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  return {
    extender: new PublicKey(data.slice(0, 32)).toBase58(),
    recipient: new PublicKey(data.slice(32, 64)).toBase58(),
    extension_type: data[64],
    bond_lamports: view.getBigUint64(65, true),
    created_at: view.getBigInt64(73, true),
    bump: data[81],
  };
}

/**
 * Extension-type discriminant → name. Mirrors the taxonomy in
 * `programs/extension-registry/src/constants.rs`. MVP records only
 * CapabilityValidation (0); the rest are reserved for weaker-verifiability
 * handling. Returns `null` for unknown discriminants.
 */
export function resolveExtensionType(extensionType: number): string | null {
  switch (extensionType) {
    case 0:
      return "capability_validation";
    case 1:
      return "mentorship";
    case 2:
      return "sponsorship";
    case 3:
      return "validation";
    default:
      return null;
  }
}
